src/java/org/apache/cassandra/index/sai/utils/RangeIterator.java

```diff
@@ -271,16 +271,16 @@ else if (tokenCount > 0 && max.compareTo(range.getMinimum()) > 0)
                             tokenCount = Math.min(tokenCount, range.getMaxKeys());
                         else
                             tokenCount = range.getMaxKeys();
+
+                        // check if new range is disjoint with already added ranges, which means that this intersection
+                        // is not going to produce any results, so we can cleanup range storage and never added anything to it.
+                        isOverlapping &= isOverlapping(min, max, range);
```

Comment:
As an aside, the call to `isOverlapping` seems somewhat unnecessary. We call it, but we don't use the information as well as we could. On one hand, we could update the `min` to be the first PK. On the other hand, we could defer this logic until the iterator is consumed, since it's not really that different, and there is a chance `skipTo` will be called before any part of this iterator is consumed, which means the `hasNext` call was a waste.